Only allow one address question per block to enable address correction

nb1701 opened the pull request:
This prevents the user from enabling address correction for question blocks that contain more than one address question. It should be disallowed by the UI, but in case it gets through, it is also disallowed by the controller and returns a 406 if this is encountered.

## Release notes

The title of the pull request will be used as the default release notes description. If more detail is needed to communicate to partners the scope of the PR's changes, use this release notes section.

### Checklist

#### General

- [ ] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

#### User visible changes

- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Manually tested at 200% size
- [ ] Manually evaluated tab order

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/application.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Issue(s) this completes

Fixes #<issue_number>; Fixes #<issue_number>...

nb1701 commented on server/app/controllers/admin/AdminProgramBlockQuestionsController.java:
```diff
@@ -20,6 +20,7 @@
 import services.program.CantAddQuestionToBlockException;
 import services.program.IllegalPredicateOrderingException;
 import services.program.InvalidQuestionPositionException;
+import services.program.ProgramBlockDefinitionContainsInvalidQuestionException;
```
This is a verrrrrry long name.  Open to suggestions to make this more concise.  Could be used in the future for other invalid questions in a program block.

seanmalbert replied:
It is informative, but could be `ProgramQuestionDefinitionInvalidException`.

nb1701 replied:
I like that better 👍

nb1701 commented on server/app/services/program/ProgramServiceImpl.java:
```diff
@@ -788,6 +789,18 @@ public ProgramDefinition setProgramQuestionDefinitionAddressCorrectionEnabled(
           "Unexpected error: updating a non address question with address correction enabled");
     }
 
+    List<ProgramQuestionDefinition> pqdWithAddressCorrectionEnabled =
```
There's perhaps a more efficient way to do this?

seanmalbert replied:
Suggestion: 
```
if (blockDefinition.programQuestionDefinitions().stream()
        .filter(
            question ->
                question.addressCorrectionEnabled() && question.id() != questionDefinitionId)
        .findAny()
        .isPresent()) {
      throw new ProgramBlockDefinitionContainsInvalidQuestionException(
          programId, blockDefinitionId, questionDefinitionId);
    }
```

Or use what you have in ProgramBlockEditView. Recommend creating a method in BlockDefinition, similar to `hasAddress` to make the check reusable.

nb1701 replied:
Actually, I don't think I actually need to collect it to a list anymore.  I think i can use the same logic as in the other place.  This was an experiment.

nb1701 replied:
Sorry, didn't reload the page and see your comment.  I'll take a look.

nb1701 replied:
Great idea, moved it into BlockDefinition.

nb1701 commented on server/app/services/program/BlockDefinition.java:
```diff
@@ -128,6 +128,19 @@ public boolean hasAddress() {
+   * the given question ID that you are trying to enable it on.
+   */
+  @JsonIgnore
+  public boolean hasAddressCorrectionEnabledOnDifferentQuestion(long questionDefinitionId) {
```
This is, once again, quite wordy.  Open to suggestions to making it more concise.

bion replied:
lgtm

nb1701 commented on server/app/services/program/ProgramServiceImpl.java:
```diff
@@ -788,6 +788,11 @@ public ProgramDefinition setProgramQuestionDefinitionAddressCorrectionEnabled(
     }
 
+    if (blockDefinition.hasAddressCorrectionEnabledOnDifferentQuestion(questionDefinitionId)) {
+      throw new ProgramQuestionDefinitionInvalidException(
```
@bion I noticed there was a recent changed that uses `BadRequestException` for something like this.  Do you think I should just use that here rather than create a new exception?

bion replied:
If there's a chance that the exceptional situation should be handled some other way than returning a 400 then go ahead and make a specific checked exception type for it, otherwise feel free to throw the `BadRequestException`.

A good thing to know about with error handling is this thing: https://github.com/civiform/civiform/blob/main/server/app/controllers/ErrorHandler.java

nb1701 replied:
I forgot I'm catching this specifically in AdminProgramBlockQuestionsController so maybe on second thought, I'll leave it.

nb1701 commented:
Added a browser test and a unit test for the BlockDefinition method I added.  Could possibly use unit tests for the ProgramServiceImpl and AdminProgramBlockQuestionsController to ensure it is throwing the right exception?  Not sure how important that is.

nb1701 commented:
Added another test
